src/parser/type_checking.c

```diff
@@ -13041,6 +13185,263 @@ pt_eval_function_type (PARSER_CONTEXT * parser, PT_NODE * node)
 	}
       break;
 
+    case F_JSON_OBJECT:
+      {
+	PT_TYPE_ENUM supported_key_types[] = { PT_TYPE_CHAR, PT_TYPE_MAYBE };
+	PT_TYPE_ENUM supported_value_types[] =
+	  { PT_TYPE_CHAR, PT_TYPE_INTEGER, PT_TYPE_DOUBLE, PT_TYPE_NUMERIC, PT_TYPE_JSON,
+	  PT_TYPE_MAYBE
+	};
+	PT_TYPE_ENUM unsupported_type;
+	unsigned int num_bad = 0, len, i, found_supported = 0;
+	int supported_value_types_len = sizeof (supported_value_types) / sizeof (supported_value_types[0]);
+	int supported_key_types_len = sizeof (supported_key_types) / sizeof (supported_key_types[0]);
+
+	PT_NODE *arg = arg_list;
+	PT_TYPE_ENUM *current_types = supported_value_types;
+
+	while (arg)
+	  {
+	    if (current_types == supported_key_types)
```

Comment:
What about using a counter variable to handle key-value list?
If I'm correct, it can be rewritten as 
```
if (i % 2 == 0)
  // key
else
  // value
```

It is easier for me to understand :)

Reply:
please also change F_JSON_INSERT case :)